[FLINK-8961][tests] Port JobRetrievalITCase to flip6

zentol opened the pull request:
## What is the purpose of the change

This PR ports the `JobRetrievalITCase` to flip6. The existing test was renamed to `LegacyJobRetrievalITCase`, and a ported copy was added.

The job retrieval works differently hence why we had to explicitly split the tests.

## Brief change log

* add utility method to `ExceptionUtils` for finding a `Throwable` matching a `Predicate`
* rename `JobRetrievalITCase` to `LegacyJobRetrievalITCase`
* add ported `JobRetrievalITCase`

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/example/client/JobRetrievalITCase.java:
```diff
@@ -21,54 +21,69 @@
+			.setNumSlotsPerTaskManager(4)
+			.build();
+		CLUSTER = new MiniCluster(clusterConfiguration);
+		CLUSTER.start();
```
We could also use a `@ClassRule` where we set the `MiniClusterType` to `FLIP6`.

zentol replied:
We can only do that if we don't want to rely on `MiniCluster#getRestAddress` for creating the client config.

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/example/client/JobRetrievalITCase.java:
```diff
@@ -21,54 +21,69 @@
+		clientConfig.setInteger(RestOptions.RETRY_MAX_ATTEMPTS, 0);
+		clientConfig.setLong(RestOptions.RETRY_DELAY, 0);
+
+		CLIENT = new RestClusterClient<>(
```
Do we want to explicitly test the `RestClusterClient`?

zentol replied:
The existing test explicitly used a `StandaloneClusterClient` for which the `RestClusterClient` is the current counter-part.

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/example/client/JobRetrievalITCase.java:
```diff
@@ -21,54 +21,69 @@
+		final Configuration clientConfig = new Configuration();
+		clientConfig.setString(JobManagerOptions.ADDRESS, "localhost");
+		clientConfig.setInteger(RestOptions.RETRY_MAX_ATTEMPTS, 0);
+		clientConfig.setLong(RestOptions.RETRY_DELAY, 0);
```
Maybe we should add a `getClusterConfiguration` method to the `MiniCluster` or `MiniClusterResource` which returns a configuration with which one can connect to the cluster.

zentol replied:
For Flip6 creating this config would be easy (just derive the host/port from `MiniCluster#getRestAddress`). I'm not sure what a `StandaloneClusterClient` would require though :/

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/example/client/JobRetrievalITCase.java:
```diff
@@ -80,42 +95,38 @@ public void testJobRetrieval() throws Exception {
-
 		// acquire the lock to make sure that the job cannot complete until the job client
 		// has been attached in resumingThread
 		lock.acquire();
```
The locking should no longer be needed in Flip-6 mode. The job result will also be available after the job has finished.

tillrohrmann replied:
Just tested it and it worked.

zentol replied:
That would no longer verify that you can connect to a _running_  job though.

tillrohrmann replied:
In Flip-6 there is no such concept as attaching to a running job anymore. 

What attaching effectively did in the legacy code, was to tell the `JobManager` that there is again a client waiting to be notified about the job result.

In Flip-6 the job result will be stored on the `Dispatcher` also if there is no client waiting to be notified. This allows the client to request it at any time (modulo some cache size cleaning conditions).

zentol replied:
But there is still a difference between asking for the result when the job is finished (where you get it right away) versus when the job running (where you have to poll).

The test should cover both.

tillrohrmann replied:
Code-wise it doesn't really make a difference for Flip-6. Given that the job takes a little bit of time, then we should fuzzily cover both cases when removing the lock, though.

zentol commented:
@tillrohrmann I've rebased and update the PR.

I've added a `getClientConfiguration()` to the `MiniClusterResource`, and converted the MiniCluster used in the test to the `MiniClusterResource`.

I have not removed the locks though. There is indeed no difference _currently_ between fetching the result for running and finished jobs, but we also want to make sure it stays this way in the future.

zentol commented:
merging.
